pkg/controller/endpoint/endpoints_controller.go

```diff
@@ -493,77 +494,85 @@ func (e *Controller) syncService(key string) error {
 	}
 	subsets = endpoints.RepackSubsets(subsets)
 
-	// See if there's actually an update here.
-	currentEndpoints, err := e.endpointsLister.Endpoints(service.Namespace).Get(service.Name)
-	if err != nil {
-		if errors.IsNotFound(err) {
-			currentEndpoints = &v1.Endpoints{
-				ObjectMeta: metav1.ObjectMeta{
-					Name:   service.Name,
-					Labels: service.Labels,
-				},
+	var createEndpoints bool
+	var newEndpoints *v1.Endpoints
+
+	// Retry on conflict
+	err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
```

Comment:
I'm not entirely sure we want `RetryOnConflict`. Let's step back, the original problem is that we're seeing a ton of events about `FailedToUpdateEndpoint` which are generated in https://github.com/kubernetes/kubernetes/blob/c964fe6001376c8ad209656175bd3c4087ce5d71/pkg/controller/endpoint/endpoints_controller.go#L584. 

It seems like something, which can be either an external actor that's racing with this controller or this controller itself, is modifying the endpoints in parallel. There are two possible approaches possible:
1. The standard controller approach, iow. fail, return error, re-queue and re-try.
2. `RetryOnConflict` which is used in some controllers (StatefulSet, for example), but only where the Update must succeed in a reasonably short time, and we can't afford to wait for full cycle from 1.

I'm not an expert in endpoints controller and the decision between 1 and 2 should be coming from someone more familiar like @aojea or other networking folks. What I can add what we do with workload controllers is we don't generate events when status update fails, at most we'll log these, but here again endpoints controller is different from other controllers :-)

Reply:
I'm fine with not generating events for Conflict errors on update
```
if !errors.IsConflict {
      generateEvent
}
``` 
@robscott what do you think?

Reply:
I agree with @soltysh here. Also the controller loops on short interval. It would be better to move to other services that are ready get their endpoints than fighting whatever cause the conflict (that shouldn't happen in the first place). Also if we have to retry then we shouldn't retry the entire code block. Just committing the changes to api-server.